### PR TITLE
feat: restructure and clean up streaming async sampling (#1130)

### PR DESCRIPTION
**Pull request by Freder-chen**

### CODE REVIEW

**Comment by hijkzzz** [comment]
TODO: I have an idea to simplify the code: for the AsyncLLM or agent scenario, we can just support streaming sampling. This way, we can remove a lot of code, and I think the code can be further simplified.

**Comment by Freder-chen** [comment]
**Pretty burned out after this one, but everything is finally in place. 🔥**

- Rebuilt sampling around `RemoteSampleGenerator` and Ray `RolloutWorker` to drive vLLM/agent/remote-RM rollouts, with dynamic filtering, pass-rate reporting, and vLLM sleep/wake handling.
- Refactored PPO trainers (sync + async) onto the new sampler: cleaner logging (wandb/TensorBoard), resumable dataloader state, dynamic batch balancing, and async buffer/signal flow control so streaming generation and weight updates don’t collide.
- Simplified experience making into a single path: unified reward/advantage computation (reward model or precomputed rewards), overlong-response penalties, group reward stats, and normalized advantages.
- Consolidated vLLM actors into one async engine (logprobs mode, collective sync) plus rollout worker creation helpers; expanded CLI/config for async queue sizing, remote RM/agent hooks, and dynamic filtering, with a new streaming training script and refreshed agent examples.

File: openrlhf/trainer/ray/rollout_worker.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+        agent_func_path: Optional[str] = None,
+        remote_rm_url: Optional[str] = None,
+        remote_rm_batch_size: Optional[int] = 1,
+        max_tasks: Optional[int] = None,
+    ):
+        # Choose executor priority: Agent > Remote RM > Plain generation.
+        if agent_func_path:
+            self.executor = AgentExecutor(agent_func_path=agent_func_path)
+        elif remote_rm_url:
+            self.executor = RewardedGenerationExecutor(remote_rm_url, remote_rm_batch_size)
```
Comment (hijkzzz) [inline openrlhf/trainer/ray/rollout_worker.py]: RLVR? RemoteRewar xxxx

File: openrlhf/trainer/ray/rollout_worker.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        else:
+            self.executor = GenerationExecutor()
```
Comment (hijkzzz) [inline openrlhf/trainer/ray/rollout_worker.py]: RewardModel

File: openrlhf/trainer/ray/rollout_worker.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+from types import SimpleNamespace
+from typing import Any, Awaitable, Callable, Optional
+
+import ray
+
+from openrlhf.utils.agent import AgentExecutorBase
+from openrlhf.utils.remote_rm_utils import RemoteRewardModel
+
+
+class RolloutExecutorBase:
```
Comment (hijkzzz) [inline openrlhf/trainer/ray/rollout_worker.py]: Rename to xxx samplegenerator
No executor

File: openrlhf/trainer/ppo_utils/sample_maker.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
+import uuid
```
Comment (hijkzzz) [inline openrlhf/trainer/ppo_utils/sample_maker.py]: 不要封装

File: openrlhf/trainer/ppo_utils/filter_hooks.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
+from typing import List, Optional
+
+from openrlhf.trainer.ppo_utils.experience_maker import Experience
+from openrlhf.utils.logging_utils import init_logger
+
+logger = init_logger(__name__)
+
+
+class FilterHookBase:
```
Comment (hijkzzz) [inline openrlhf/trainer/ppo_utils/filter_hooks.py]: 不要封装

File: openrlhf/trainer/ppo_trainer_async.py [hunk 6]
```diff
@@ -1,28 +1,59 @@
...
 from openrlhf.trainer.ray.launcher import RayActorGroup
 from openrlhf.utils.deepspeed import DeepspeedStrategy
 from openrlhf.utils.logging_utils import init_logger
+from openrlhf.utils.utils import get_tokenizer
 
 logger = init_logger(__name__)
 
 
+@ray.remote(num_cpus=0)
+class BufferActor:
```
Comment (hijkzzz) [inline openrlhf/trainer/ppo_trainer_async.py]: 不要封装

**Comment by Freder-chen** [comment]
Potential follow-up features (to discuss first; can be implemented in separate PRs later)

1) Public interface / backward compatibility: `remote_rm_url` vs `agent_func_path`
   We may want to refactor the exposed CLI/API around `remote_rm_url` and `agent_func_path` to make the “reward / rollout backend” selection explicit and less ambiguous. This is likely to impact existing users and configs, so it should be discussed up front (migration path, deprecation window, precedence rules, and validation behavior).

2) Ray CPU budgeting for CPU-heavy agent environments
   If the agent environment/tools require substantial CPU, we’ll likely need to refactor Ray resource allocation so rollout workers can actually obtain more CPU (e.g., configurable CPU per rollout worker / per GPU bundle, and possibly decoupling env/tool execution into separate actors). This can be kept as an opt-in knob first, then refined.

3) Evaluation parity (currently missing)
   Evaluation is not wired in yet for the new async pipeline. We should re-introduce periodic eval (keeping CLI semantics and metrics parity), and in async mode ideally run eval in a dedicated actor/group so it doesn’t stall sampling/training.

These are not blockers for this PR, but they are likely necessary for stability/usability, so I’d like to align on the direction before starting follow-up PRs.

File: openrlhf/trainer/ppo_trainer_async.py [hunk 7]
```diff
@@ -249,104 +206,79 @@ def __init__(
...
+            actor_model_group=actor_model_group,
+            critic_model_group=critic_model_group,
+            reward_model_group=reward_model_group,
+            reference_model_group=reference_model_group,
+            vllm_engines=vllm_engines,
             signal_actor=self.signal_actor,
-            remote_reward_model=self.remote_reward_model,
-            **generate_kwargs,
+            rollout_queue=self.rollout_queue,
+            rollout_slots=self.rollout_slots,
```
Comment (Freder-chen) [inline openrlhf/trainer/ppo_trainer_async.py:237]: delete

File: openrlhf/trainer/ppo_utils/experience_maker.py [hunk 8]
```diff
@@ -205,251 +210,297 @@ def concat_experiences(experiences_list: List["Experience"], pad_token_id) -> "E
...
+        self, dataloader_iter, num_prompts: int, **generate_kwargs
+    ) -> Tuple[List[Experience], int, bool]:
+        """Generate a batch of Experiences with optional reward filtering."""
+        prompts_consumed = 0
+        prompts, labels, exhausted = _collect_prompt_batch(dataloader_iter, num_prompts)
+        # Stop early if the prompt source is fully consumed.
+        if exhausted:
+            return [], prompts_consumed, exhausted
+
+        pending_refs = self._dispatch_prompts(prompts, labels, **generate_kwargs)
```
Comment (Freder-chen) [inline openrlhf/trainer/ppo_utils/experience_maker.py]: _dispatch_prompt_to_vllm

File: openrlhf/trainer/ppo_utils/experience_maker.py [hunk 9]
```diff
@@ -205,251 +210,297 @@ def concat_experiences(experiences_list: List["Experience"], pad_token_id) -> "E
...
+
+        accepted_experiences: List[Experience] = []
+        pbar = tqdm(range(num_prompts), desc="Generate samples")
+
+        while pending_refs:
+            ready_refs, pending_refs = ray.wait(pending_refs, num_returns=1, timeout=10.0)
+            for ref in ready_refs:
+                # Build Experience objects for each vLLM response returned from this worker.
+                experiences = [
+                    self._process_response_into_experience(response, **generate_kwargs) for response in ray.get(ref)
```
Comment (Freder-chen) [inline openrlhf/trainer/ppo_utils/experience_maker.py:325]: ?

File: openrlhf/trainer/ray/vllm_engine.py [hunk 10]
```diff
@@ -1,112 +1,177 @@
...
+    MultiTurnExecutor,
+    SingleTurnExecutor,
+    SingleTurnRewardedExecutor,
+)
 
 from .utils import get_bundle_indices, ray_noset_visible_devices
 
-logger = init_logger(__name__)
 
+class BaseLLMRayActor:
```
Comment (Freder-chen) [inline openrlhf/trainer/ray/vllm_engine.py]: BaseVLLMRayActor

File: openrlhf/trainer/ray/vllm_engine.py [hunk 11]
```diff
@@ -1,112 +1,177 @@
...
-        from packaging import version
+            os.environ["RAY_ADDRESS"] = global_worker.gcs_client.address
 
-        if version.parse(vllm.__version__) >= version.parse("0.9.0"):
-            os.environ["VLLM_ALLOW_INSECURE_SERIALIZATION"] = "1"
+        os.environ["VLLM_USE_V1"] = "1"
 
 
 @ray.remote
 class LLMRayActor(BaseLLMRayActor):
```
Comment (Freder-chen) [inline openrlhf/trainer/ray/vllm_engine.py]: merge base actor

File: openrlhf/utils/rollout.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+
+import aiohttp
+
+from openrlhf.utils.agent import AgentExecutorBase
+from openrlhf.utils.logging_utils import init_logger
+
+logger = init_logger(__name__)
+
+
+class SingleTurnExecutor:
```
Comment (Freder-chen) [inline openrlhf/utils/rollout.py]: SingleTurnAgentExecutor
